src/arizona_template_compiler.erl

```diff
@@ -0,0 +1,646 @@
+-module(arizona_template_compiler).
+-moduledoc false.
+
+%% --------------------------------------------------------------------
+%% API function exports
+%% --------------------------------------------------------------------
+
+-export([compile/3]).
+
+%
+
+-ignore_xref([compile/3]).
+
+%% --------------------------------------------------------------------
+%% Types (and their exports)
+%% --------------------------------------------------------------------
+
+-record(state, {
+    module :: module(),
+    function :: atom(),
+    macros :: macros(),
+    attributes :: [arizona_template_parser:attribute()],
+    index :: non_neg_integer(),
+    path :: [non_neg_integer()]
+}).
+
+-type macros() :: #{atom() := arizona_html:safe_type()}.
+-export_type([macros/0]).
+
+-opaque changeable() :: {expr, expr()} | {block, block()}.
+-export_type([changeable/0]).
+
+-opaque expr() :: #{
+    id := [non_neg_integer()],
+    function := fun((Assigns :: map()) -> arizona_html:safe_type()),
+    vars := [atom()]
+}.
+-export_type([expr/0]).
+
+-opaque block() :: #{
+    id => [non_neg_integer()],
+    module => module(),
+    function => atom(),
+    static => [binary()],
+    changeable => #{non_neg_integer() := changeable()},
+    changeable_vars => #{atom() := [[non_neg_integer()]]},
+    changeable_indexes => [non_neg_integer()],
+    norm_assigns => #{atom() := {text, binary()} | {expr, expr()}},
```

Comment:
Hm this seems to be not true. We only have expressions here.
```suggestion
    norm_assigns => #{atom() := expr()},
```

Reply:
[ddb06fa](https://github.com/arizona-framework/arizona/pull/164/commits/ddb06faa953bdba094e2c39b6d793baa51f6ac51)